IPython/core/guarded_eval.py

```diff
@@ -853,6 +858,116 @@ def dummy_function(*args, **kwargs):
     return None
 
 
+def _extract_init_attributes(class_node: ast.ClassDef, context: EvaluationContext):
+    """Extract attribute assignments from __init__ method.
+
+    Looks for patterns like:
+        self.attr = value
+        self.attr: Type = value
+
+    And infers their types by evaluating the assigned values.
+
+    Returns dictionary mapping attribute names to their inferred Duck types
+    """
+    attributes = {}
+    init_method = None
+    for node in class_node.body:
+        if isinstance(node, ast.FunctionDef) and node.name == "__init__":
+            init_method = node
+            break
+
+    if not init_method:
+        return attributes
+
+    for stmt in init_method.body:
```

Comment:
I think we may not need `_extract_init_attributes` no `_extract_variables_from_module`. In particular we may _not_ need to:
- iterate over the class body for the second time (instead we would just handle that in `eval_node → func` case); we already have `transient_locals` which represent the locals of a class. `transient_locals` were a quick fix but a better implementation would be tracking scopes explicitly and each time we enter a class or function definition we would push a new scope onto a list, each time we exit we would pop it; the difference is that it would allow to correctly access variables defined in parent scopes.
- look for `__init__` specifically

Reply:
> each time we enter a class or function definition we would push a new scope onto a list, each time we exit we would pop it;

which is kind of trivial since we are calling `eval_node` recursively, this gets handled by itself if we are making copies of context when entering `eval_node` for class elements